pkg/metrics/metrics.go

```diff
@@ -0,0 +1,270 @@
+		prometheus.CounterOpts{
+			Name: "webhook_request_reject_count",
+			Help: "Total number of rejected validating and mutating webhook calls",
+		},
+		[]string{"type", "group", "version", "kind", "namespace", "name"},
+	)
+	webhookRequestPanicCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "webhook_request_panic_count",
+			Help: "Total number of panics during validating and mutating webhook calls",
+		},
+		[]string{"type", "group", "version", "kind", "namespace", "name"},
+	)
+
+	// Controller metrics vectors have the following attributes
+	// group: The target resource group of the webhook, e.g., risingwave.risingwavelabs.com
+	// version: The target API version, e.g., v1alpha1
+	// kind: The target API kind, e.g., risingwave, risingwavepodtemplate
+	// namespace: The namespace of the object, e.g., default
+	// name: The name of the object
+	// TODO: verb: The verb (action) on the object which triggers the webhook, the value should be one of "create", "update", and "delete".
+	controllerReconcileCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_reconcile_count",
+			Help: "Total number of reconciles",
+		},
+		[]string{"group", "version", "kind", "namespace", "name"},
+	)
+	controllerReconcileRequeueCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_reconcile_requeue_count",
+			Help: "Total number of requeue",
+		},
+		[]string{"group", "version", "kind", "namespace", "name"},
+	)
+	controllerReconcileRequeueErrorCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_reconcile_error_count",
+			Help: "Total number of requeue errors",
+		},
+		[]string{"group", "version", "kind", "namespace", "name"},
+	)
+
+	controllerReconcileRequeueAfter = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "controller_reconcile_requeue_after",
+			Help:    "Delay of last delayed requeue in ms",
+			Buckets: []float64{1, 10, 50, 100, 500, 1000}, // wait time in ms
+		},
+		[]string{"group", "version", "kind", "namespace", "name"},
+	)
+	controllerReconcileDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "controller_reconcile_duration",
+		Help: "Length of time per reconciliation per controller",
+		Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0,
+			1.25, 1.5, 1.75, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30, 40, 50, 60},
+	}, []string{"controller", "group", "version", "kind", "namespace", "name"})
+	controllerReconcilePanicCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "controller_reconcile_panic_count",
+			Help: "Total number of reconcile panics",
+		},
+		[]string{"group", "version", "kind", "namespace", "name"},
+	)
+)
+
+// defaultNamespaceName applies defaults in case the attributes of nn are undefined.
+func defaultNamespaceName(nn *types.NamespacedName) types.NamespacedName {
```

Comment:
I guess we don't have to specify the default values since the empty string "" is already an undefined value.